File: scripts/fetch_meta.py
import requests
from bs4 import BeautifulSoup
import json
import re

def fetch_meta():
    url = "https://play.limitlesstcg.com/decks?game=POCKET"
    headers = {'User-Agent': 'Mozilla/5.0'}
    
    try:
        response = requests.get(url, headers=headers, timeout=30)
        soup = BeautifulSoup(response.text, 'html.parser')
        
        # Find deck cards
        deck_cards = soup.find_all('div', class_='deck-card')
        
        decks = []
        for i, card in enumerate(deck_cards[:50]):
            try:
                name_elem = card.find('a')
                stats_elem = card.find_all('span')
                
                name = name_elem.text.strip() if name_elem else "Unknown"
                
                # Extract stats from text
                card_text = card.get_text()
                
                # Parse usage rate
                usage_match = re.search(r'([\d.]+)%', card_text)
                usage = usage_match.group(1) + "%" if usage_match else "0%"
                
                # Parse win rate
                win_match = re.search(r'([\d.]+)%.*?(\d+)\s*-\s*(\d+)', card_text)
                if win_match:
                    win_rate = win_match.group(1) + "%"
                    wins = int(win_match.group(2))
                    losses = int(win_match.group(3))
                    record = f"{wins} - {losses}"
                else:
                    win_rate = "0%"
                    record = "0 - 0"
                
                # Calculate adjusted win rate (Bayesian smoothing)
                prior_games = 25
                prior_wins = 12
                if wins + losses > 0:
                    adjusted = (wins + prior_wins) / (wins + losses + prior_games)
                    adjusted_win_rate = f"{adjusted*100:.2f}%"
                else:
                    adjusted_win_rate = "0%"
                
                deck = {
                    "rank": i + 1,
                    "name": name,
                    "usage_rate": usage,
                    "win_rate": win_rate,
                    "record": record,
                    "adjusted_win_rate": adjusted_win_rate,
                    "sample_size": wins + losses
                }
                decks.append(deck)
                
            except Exception as e:
                continue
        
        # Save to file
        result = {
            "update_time": "2026-03-14 12:00",
            "top_decks": decks
        }
        
        with open('/home/node/.openclaw/workspace/memory/ptcgp_meta.json', 'w', encoding='utf-8') as f:
            json.dump(result, f, ensure_ascii=False, indent=2)
        
        print(f"Updated {len(decks)} decks")
        return decks
        
    except Exception as e:
        print(f"Error: {e}")
        return []

if __name__ == "__main__":
    fetch_meta()

File: scripts/fetch_meta.mjs
import fetch from 'node-fetch';
import * as cheerio from 'cheerio';
import { writeFileSync } from 'fs';

async function fetchMeta() {
  const url = "https://play.limitlesstcg.com/decks?game=POCKET";
  
  try {
    const response = await fetch(url, {
      headers: {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36'
      },
      timeout: 30000
    });
    
    const html = await response.text();
    const $ = cheerio.load(html);
    
    const decks = [];
    
    // Try multiple selectors for Limitless
    const deckSelectors = [
      '.deck-card',
      '[class*="deck"]',
      '.tournament-deck',
      'article',
      '.card'
    ];
    
    // Look for deck entries - try to find any ranked list
    const allElements = $('*');
    let deckCount = 0;
    
    // Try finding by text patterns - deck names usually contain "ex" or specific patterns
    $('a[href*="deck"]').each((i, el) => {
      if (deckCount >= 20) return;
      const name = $(el).text().trim();
      const parent = $(el).parent();
      const grandparent = parent.parent();
      const container = grandparent.parent();
      const text = container.text();
      
      if (name && name.length > 3 && name.length < 60) {
        // Extract percentages
        const usageMatch = text.match(/([\d.]+)%/);
        const usage = usageMatch ? usageMatch[1] + '%' : 'N/A';
        
        // Extract win/loss
        const recordMatch = text.match(/(\d+)\s*-\s*(\d+)/);
        let wins = 0, losses = 0, winRate = 'N/A';
        if (recordMatch) {
          wins = parseInt(recordMatch[1]);
          losses = parseInt(recordMatch[2]);
          const total = wins + losses;
          if (total > 0) winRate = ((wins / total) * 100).toFixed(2) + '%';
        }
        
        // Bayesian adjusted
        let adjWinRate = 'N/A';
        if (wins + losses > 0) {
          const priorGames = 25, priorWins = 12;
          adjWinRate = ((wins + priorWins) / (wins + losses + priorGames) * 100).toFixed(2) + '%';
        }
        
        // Check if it's a Pocket deck (has ex, Mega, etc.)
        if (name.includes('ex') || name.includes('Mega') || name.includes('Pika') || name.length < 30) {
          decks.push({
            rank: deckCount + 1,
            name,
            usage_rate: usage,
            win_rate: winRate,
            record: `${wins} - ${losses}`,
            adjusted_win_rate: adjWinRate,
            sample_size: wins + losses
          });
          deckCount++;
        }
      }
    });
    
    // Fallback: try to get from JSON data embedded in page
    if (decks.length === 0) {
      const scriptMatch = html.match(/window\.__INITIAL_STATE__\s*=\s*({.+?});/s);
      const jsonMatch = html.match(/type="application\/json">(.+?)</);
      
      if (jsonMatch) {
        try {
          const data = JSON.parse(jsonMatch[1]);
          // Try to extract deck data
          console.log('Found JSON data in page');
        } catch (e) {}
      }
    }
    
    // Save raw HTML for debugging
    const workspacePath = '/home/node/.openclaw/workspace-celesteela';
    writeFileSync(`${workspacePath}/memory/ptcgp_raw_latest.html`, html);
    
    const result = {
      update_time: new Date().toISOString(),
      top_decks: decks
    };
    
    writeFileSync(`${workspacePath}/memory/ptcgp_meta_latest.json`, JSON.stringify(result, null, 2));
    
    console.log(`Found ${decks.length} decks`);
    console.log(JSON.stringify(decks.slice(0, 5), null, 2));
    
    return decks;
    
  } catch (e) {
    console.error('Error:', e.message);
    return [];
  }
}

fetchMeta();
